File: rust/expr.rs
enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
}

enum Expr<'a> {
    AtomicExpr { val: isize },
    BinaryExpr {
        op: BinOp,
        lhs: &'a Expr<'a>,
        rhs: &'a Expr<'a>,
    },
}

impl<'a> Expr<'a> {
    fn eval(&self) -> isize {
        use Expr::*;

        match self {
            AtomicExpr { val } => *val,
            BinaryExpr { op, lhs, rhs } => match op {
                BinOp::Add => lhs.eval() + rhs.eval(),
                BinOp::Sub => lhs.eval() - rhs.eval(),
                BinOp::Mul => lhs.eval() * rhs.eval(),
                BinOp::Div => lhs.eval() / rhs.eval(),
            }
        }
    }
}

fn main() {
    let expr = Expr::BinaryExpr {
        op: BinOp::Add,
        lhs: &Expr::AtomicExpr { val: 5 },
        rhs: &Expr::BinaryExpr {
            op: BinOp::Mul,
            lhs: &Expr::AtomicExpr { val: 12 },
            rhs: &Expr::AtomicExpr { val: 4 },
        },
    };

    println!("{}", expr.eval());
}

File: python/expr.py
from enum import Enum

class BinOp(Enum):
    ADD = 0
    SUB = 1
    MUL = 2
    DIV = 3

def eval(expr):
    if (type(expr) == int or type(expr) == float):
        return expr
    elif (type(expr)):
        (op, lhs, rhs) = expr
        op = {
                BinOp.ADD: lambda a, b: a + b,
                BinOp.SUB: lambda a, b: a - b,
                BinOp.MUL: lambda a, b: a * b,
                BinOp.DIV: lambda a, b: a / b,
        }[op]
        return op(eval(lhs), eval(rhs))
    else:
        raise Exception()

e = (BinOp.ADD, 3, (BinOp.MUL, 4, (BinOp.ADD, 2, 3)))
print(eval(e))
